cmd/agent/gui/checks.go

```diff
@@ -185,56 +193,81 @@ func setCheckConfigFile(w http.ResponseWriter, r *http.Request) {
-	e = ioutil.WriteFile(path, data, 0644)
+	// Attempt to write new configs to custom checks directory
+	path := filepath.Join(config.Datadog.GetString("confd_path"), getPathToFile(fileName))
+	e = ioutil.WriteFile(path, data, 0600)
+
+	// If the write didn't work, try writing to the default checks directory
+	if e != nil && strings.Contains(e.Error(), "no such file or directory") {
+		path = filepath.Join(common.GetDistPath(), "conf.d", getPathToFile(fileName))
+		e = ioutil.WriteFile(path, data, 0600)
+	}
+
 	if e != nil {
-		w.Write([]byte("Error writing to " + fileName + ": " + e.Error()))
+		w.Write([]byte("Error saving config file: " + e.Error()))
+		log.Debug("Error saving config file: " + e.Error())
 		return
 	}
 
 	log.Infof("Successfully wrote new " + fileName + " config file.")
 	w.Write([]byte("Success"))
 }
 
-// Sends a list containing the names of all the specified type of files
-func listFiles(w http.ResponseWriter, r *http.Request) {
-	// Get the directories where these files might reside
-	fileType := mux.Vars(r)["fileType"]
-	paths := getAllPaths(fileType)
+// Sends a list containing the names of all the checks
+func listChecks(w http.ResponseWriter, r *http.Request) {
+	filenames := []string{}
+	for _, path := range checkPaths {
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			continue
+		}
+
+		for _, file := range files {
+			if ext := filepath.Ext(file.Name()); ext == ".py" && file.Mode().IsRegular() {
+				filenames = append(filenames, file.Name())
+			}
+		}
+	}
+
+	if len(filenames) == 0 {
+		w.Write([]byte("No check (.py) files found."))
+		return
+	}
+
+	res, _ := json.Marshal(filenames)
+	w.Header().Set("Content-Type", "application/json")
+	w.Write([]byte(res))
+}
 
-	// Read all the directories
+// Sends a list containing the names of all the config files
+func listConfigs(w http.ResponseWriter, r *http.Request) {
 	filenames := []string{}
-	lookup := make(map[string]bool)
-	for _, path := range paths {
-		fs, e := readDirectory(path)
+	for _, path := range configPaths {
+		files, e := readConfDir(path)
+
 		if e == nil {
-			sort.Strings(fs)
-			for _, name := range fs {
-				// Only include each file name once (could be in multiple locations),
-				// & if a default config is found but a non-default version exists, don't include
-				// the default one (Note that the non-default directory is read first)
-				trimmed := name
-				if i := strings.Index(name, ".default"); i != -1 {
-					trimmed = name[:i]
-				}
-				if _, exists := lookup[trimmed]; !exists {
-					filenames = append(filenames, name)
-					lookup[trimmed] = true
+			// If a default config is found but a non-default version exists, ignore default
+			sort.Strings(files)
+			lookup := make(map[string]bool)
+			for _, file := range files {
+				checkName := getCheckName(file)
+
+				if ext := filepath.Ext(file); ext == ".default" {
+					if _, exists := lookup[checkName]; !exists {
+						filenames = append(filenames, file)
```

Comment:
we could invert the condition and `continue` in the if. This would avoid duplicating those 2 lines.